### PR TITLE
Create is valid email address algorithm

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: strings/is_valid_email_address.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    ("example@s.example", True),
+    ("Abc.example.com", False),
+    ("A@b@c@example.com", False),
+    ("abc@example..com", False),
+    ("a(c)d,e:f;g<h>i[j\\k]l@example.com", False),
+    (
+        "12345678901234567890123456789012345678901234567890123456789012345@example.com",
+        False,
+    ),
+    ("i.like.underscores@but_its_not_allowed_in_this_part", False),
```
Comment (madhavgaba) [inline strings/is_valid_email_address.py:34]: can you add for empty string?

File: strings/is_valid_email_address.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        for char in local_part
+    ):
+        return False
+
+    # (4.) Validate the placement of "." characters
+    if (
+        local_part.startswith(".")
+        or local_part.endswith(".")
+        or re.search(r"\.\.+", local_part)
+    ):
```
Comment (tianyizheng02) [inline strings/is_valid_email_address.py:96]:
```suggestion
    # (4.) Validate the placement of "." characters in the local-part
    if (
        local_part.startswith(".")
        or local_part.endswith(".")
        or ".." in local_part
    ):
```
No regex necessary

File: strings/is_valid_email_address.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    # (5.) Validate the characters in the domain
+    if any(char not in string.ascii_letters + string.digits + ".-" for char in domain):
+        return False
+
+    # (6.) Validate the placement of "-" characters
+    if domain.startswith("-") or domain.endswith("."):
+        return False
+
+    # (7.) Validate the placement of "." characters
+    if domain.startswith(".") or domain.endswith(".") or re.search(r"\.\.+", domain):
```
Comment (tianyizheng02) [inline strings/is_valid_email_address.py:108]:
```suggestion
    # (7.) Validate the placement of "." characters in the domain
    if domain.startswith(".") or domain.endswith(".") or ".." in domain:
```
No regex necessary

File: strings/is_valid_email_address.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        last character, and may not have more than one "." consecutively.
+
+    Checks for the domain:
+     5. The domain may contain: upper and lowercase latin letters and digits 0 to 9
+     6. Hyphen "-", provided that it is not the first or last character
+     7. The domain may also contain a "." in any place that is not the first or
+        last character, and may not have more than one "." consecutively.
+
+    >>> for email, valid in email_tests:
+    ...     assert is_valid_email_address(email) is valid
```
Comment (tianyizheng02) [inline strings/is_valid_email_address.py:72]:
```suggestion
    ...     assert is_valid_email_address(email) == valid
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: strings/is_valid_email_address.py (hunk 1 above)
Comment (CaedenPH) [reply strings/is_valid_email_address.py:34]: Good idea!

File: strings/is_valid_email_address.py (hunk 3 above)
Comment (CaedenPH) [reply strings/is_valid_email_address.py:108]: Good point

File: strings/is_valid_email_address.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
+
+    for email, valid in email_tests:
+        is_valid = is_valid_email_address(email)
+        assert is_valid == valid, f"{email} is {is_valid}"
+        print(f"Email address {email} is {'not ' if is_valid else ''}valid")
```
Comment (tianyizheng02) [inline strings/is_valid_email_address.py:117]:
```suggestion
        print(f"Email address {email} is {'not ' if not is_valid else ''}valid")
```
Typo?

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: strings/is_valid_email_address.py (hunk 5 above)
Comment (CaedenPH) [reply strings/is_valid_email_address.py]: Yep 😁

**Review by tianyizheng02 — APPROVED** [review]
LGTM

**Pull request merged**